### PR TITLE
Add support for configuring kubectl in pipelines

### PR DESCRIPTION
**Pull request by maxlaverse**
The `kubernetes-plugin` provides a `SimpleBuildWrapper` configurable through the UI than can be used to provide `kubectl` with a `kubeconfig` filled with Jenkins credentials.

This is not the main feature of this plugin but still very useful. Sadly, this is not supported in Jenkinsfiles.

This PR would make it available through a new `withKubeConfig` function that accepts the same parameters as the UI (api-server url, ca certificate and credential id).

It also includes a few other changes:
* Embed certificates into the kube config to have only one file to clean after the build
* Change the masks used on the command to set the login credentials to prevent tokens from leaking into the logs
* Change the way the plugin retrieves credentials to be able to read scoped one

Since I didn't know if you would agree to add this feature, I didn't updated the README yet.

### CODE REVIEW

**Comment by carlossg** [comment]
Thanks, can you pos an example on how this would be used?

**Comment by maxlaverse** [comment]
Sure.
```
pipeline{
    agent any
    
    environment {
        # Mandatory - Can be scoped to a folder or a Github Org
        KUBERNETES_CREDENTIALS_ID = 'max-test-k8s'

        # Mandatory
        KUBERNETES_URL = 'https://kubernetes.api.dc.local'

        # Optional 
        KUBERNETES_CA = '''-----BEGIN CERTIFICATE-----
MIIFlDCCA3ygAwIBAgIJAK07Q+xQ0nOXMA0GCSqGSIb3DQEBCwUAMHcxCzAJBgNV
HKu8P6rt5ECg1YIT8Wu15CxHQrm1/7Kb/rTiYhibGsAKqi+h0Q6uGQ==
-----END CERTIFICATE-----
'''
    }
    
    stages {
        stage('List pods') {
            steps {
                withKubectl([caCertificate: env.KUBERNETES_CA, credentialsId: env.KUBERNETES_CREDENTIALS_ID, serverUrl: env.KUBERNETES_URL]) {
                    sh '''
                        kubectl -n max get pods
                    '''
                }
            }
        }
    }
}
```

**Comment by jmcshane** [comment]
Would this also allow the `cloud` section to be skipped and have it read from the `withKubectl ` block itself? That's the feature that I'm looking for, which is to define a pod template wrapped in a `withKubectl` block so that I can do pipeline work across multiple clusters easily, rather than having to define the cluster in my Jenkins master config file.

**Comment by maxlaverse** [comment]
I can't say @jmcshane, I haven't used the cloud section.
Feel free to run: `mvn package`, install the hpi on your Jenkins and give it a try ;)

**Comment by maxlaverse** [comment]
Hi @carlossg 
Would you be interested in that PR ? No hurry, and no hurt feelings, but if you think this is out of the scope of this plugin, I would extract that part in a separate plugin. It would be easier for us to maintain on our installations and we wouldn't have to pin the kubernetes-plugin version.

**Comment by carlossg** [comment]
Sorry for the delay, I do think it would be valuable, but would need some docs in the readme and tests (unit/integration) like the ones for the other pipeline steps.

Also, could you move the pipelines classes to pipeline/kubectl/... package?

**Comment by maxlaverse** [comment]
@carlossg Should I mock the `kubectl` calls or is do you want to bring the tool on the slave for the integration tests ?

**Comment by maxlaverse** [comment]
Hi @carlossg !
Would you have time giving me some indication how to proceed ?
I'm ready to put more work in the PR to have it merge.

**Comment by maxlaverse** [comment]
or @marvinthepa maybe ?

**Comment by carlossg** [comment]
sorry for the delay. No mocking, the test should execute kubectl to list pods for example. It should work against minikube and in my testing GKE cluster (I can help with that)

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/kubectl/KubectlBuildWrapperStepTest.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    protected static final String PASSPHRASE = "test";
+    protected static final String USERNAME = "bob";
+    protected static final String PASSWORD = "s3cr3t";
+    protected static final String CA_CERTIFICATE = "-----BEGIN CERTIFICATE-----\na-certificate\n-----END CERTIFICATE-----";
+    protected static final String SERVER_URL = "https://localhost:6443";
+
+    @Rule
+    public JenkinsRuleNonLocalhost r = new JenkinsRuleNonLocalhost();
+
+    protected String loadPipelineScript(String name) {
```
Comment (carlossg) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/kubectl/KubectlBuildWrapperStepTest.java]: you'll need to run this only when in a kubernetes environment

```
    @BeforeClass
    public static void isKubernetesConfigured() throws Exception {
        assumeKubernetes();
    }
```

**Comment by maxlaverse** [comment]
For this PR, since all the commands are only related to `kubectl config`, it doesn't even need to connect to a cluster. But if you meant using `kubectl get pods` to check if the credentials were correctly set, then yes, I would need some information about the cluster test setup and help would be appreciated. How can we work on this ? Asynchronously, or should we try to be chat about it @carlossg ?

**Comment by maxlaverse** [comment]
Hi @carlossg 
Sorry to bother you one last time.

I've always answered pretty fast to your comments, willing to make the necessary changes. I've seen you've commited many times to the repository since I opened the pull-request, so I don't think time is the issue here. 
Let's be honest: this PR is obvisouly not an important feature for the kubernetes plugin. No hurt feelings at all. I knew it from the start when I opened to pull-request and showed my hesitation in the pr description. This is an non-documented feature of the kubernetes plugin. It's stated everywhere that the main focus is about provisioning slaves on Kubernetes.

The kubernetes plugin is alive and gets updates which is great. But it has a cost to maintain a fork with that PR. Therefore I extracted the kubectl related features in a different plugin, kubectl. It's basically 6 classes:
* BasicTokenCredentialImpl (just a rename of the Openshit one)
* KubeConfigExpander
* KubectlBuildWrapper
* KubectlBuildWrapperStep
* TokenProducer
* KubectlBuildWrapperStepTest

This new plugin is simple, has tests, few dependencies and a low maintainance cost as the kubectl interface is quit stable. It has the advantages of this PR: no more leaking of tokens and the ability to use scoped credentials.

Now. This is a question for you. Is it worth asking to publish that new plugin on jenkins-ci and opening a pull-request here to make use of it ?

It would allow you to concentrate on the slaves features and give people the possibility to use the kubectl configuration through pipelines and UI without having to install the kubernetes plugin.
I believe it makes as much sense if not more than having this feature here. A bit like the `docker-commons-plugin` that is used by other Docker based plugins.

Have a nice week-end

**Comment by carlossg** [comment]
I'm really sorry, I definitely have a huge backlog. I hoped that somebody else would chime in and 👍  or comment on the PRs

I think it makes sense to have it in a separate plugin, but what we can do to avoid duplication is to create a new `kubernetes-credentials-plugin` that has those shared classes.

Does it make sense?

**Comment by maxlaverse** [comment]
Yes sure! 
Which groupId should I use?
* `org.csanchez.jenkins.plugins`
* `org.jenkinsci.plugins`
* mine ?

**Comment by carlossg** [comment]
You can use org.jenkinsci.plugins or your own, I think if you ask in IRC an admin can create the plug-in repo inside jenkins org

**Comment by maxlaverse** [comment]
Hi
I released a separate plugin `kubernetes-cli` today:
https://plugins.jenkins.io/kubernetes-cli (see also https://github.com/jenkinsci/kubernetes-cli-plugin).

It uses another plugin `kubernetes-credentials` which holds the `TokenProducer` interface but also the OpenShift implementations (see https://github.com/jenkinsci/kubernetes-credentials-plugin).

Two questions for you @carlossg:
* Is it okay to have the two OpenShift implementations in the shared plugin or do you want to keep them here ? I don't think a lot of people. That will force people to update their credentials for the new class.
* If it's okay to have those implementations in the shared plugin, do you want to keep this `BasicOpenShiftTokenCredentialImpl` ? The `plain-credentials` plugins which is widely uses already brings this functionnality and I've added support for it in the `kubernetes-cli`.

**Pull request closed without merging**